agent/src/util/device_plugin_service.rs

```diff
@@ -179,23 +180,26 @@ impl DevicePlugin for DevicePluginService {
                             .unwrap();
                 }
 
-                let resp = v1beta1::ListAndWatchResponse {
-                    devices: virtual_devices,
-                };
-
-                // Send virtual devices list back to kubelet
-                if let Err(e) = kubelet_update_sender.send(Ok(resp)).await {
-                    trace!(
-                        "list_and_watch - for Instance {} kubelet no longer receiving with error {}",
-                        dps.instance_name,
-                        e
-                    );
-                    // This means kubelet is down/has been restarted. Remove instance from instance map so
-                    // do_periodic_discovery will create a new device plugin service for this instance.
-                    dps.instance_map.lock().await.remove(&dps.instance_name);
-                    dps.server_ender_sender.clone().send(()).await.unwrap();
-                    keep_looping = false;
+                if prev_virtual_devices != virtual_devices {
```

Comment:
This assumes that the discovery handler always returns the devices in the same order. For example, the following assertion fails:
```sh
    let vec1 = vec![1,2,3];
    let vec2 = vec![2,1,3];
    assert!(vec1 == vec2);
```
You may consider using HashSets.